bundles/org.openhab.binding.miio/src/main/java/org/openhab/binding/miio/internal/handler/MiIoAbstractHandler.java

```diff
@@ -93,18 +102,35 @@
     private final Logger logger = LoggerFactory.getLogger(MiIoAbstractHandler.class);
     protected MiIoDatabaseWatchService miIoDatabaseWatchService;
 
-    public MiIoAbstractHandler(Thing thing, MiIoDatabaseWatchService miIoDatabaseWatchService) {
+    public MiIoAbstractHandler(Thing thing, MiIoDatabaseWatchService miIoDatabaseWatchService,
+            CloudConnector cloudConnector) {
         super(thing);
         this.miIoDatabaseWatchService = miIoDatabaseWatchService;
+        this.cloudConnector = cloudConnector;
     }
 
     @Override
     public abstract void handleCommand(ChannelUID channelUID, Command command);
 
+    protected boolean handleCommandsChannels(ChannelUID channelUID, Command command) {
+        if (channelUID.getId().equals(CHANNEL_COMMAND)) {
+            cmds.put(sendCommand(command.toString(), ""), command.toString());
+            return true;
+        }
+        if (channelUID.getId().equals(CHANNEL_RPC)) {
+            cmds.put(sendCommand(command.toString(), cloudServer), command.toString());
+            return true;
+        }
+        return false;
+    }
+
     @Override
     public void initialize() {
         logger.debug("Initializing Mi IO device handler '{}' with thingType {}", getThing().getUID(),
                 getThing().getThingTypeUID());
+
+        miIoScheduler = ThreadPoolManager.getScheduledPool(getThing().getUID().getAsString());
```

Comment:
You shouldn't perform shutdown of the threadpools returned by ThreadPoolManager.
Instead you should create your own ScheduledThreadPoolExecutor instance as follows:

```suggestion
        miIoScheduler = new ScheduledThreadPoolExecutor(1,
                new NamedThreadFactory(getThing().getUID().getAsString(), true));
        miIoScheduler.setExecuteExistingDelayedTasksAfterShutdownPolicy(false);
        miIoScheduler.setRemoveOnCancelPolicy(true);
```
The example above creates a scheduler with only a single thread, but you can increase this number as you see fit.

Reply:
I was very hesitant to do that as in https://github.com/openhab/openhab-core/blob/master/bundles/org.openhab.core/src/main/java/org/openhab/core/common/ThreadPoolManager.java it was made clear that bindings really should not do that.

I thought that that what the ThreadPoolManager was doing for me... if the name does not exist yet, create a new ScheduledThreadPoolExecutor. This is why I added the thingUID in there to make sure it was a unique one (for that thing) instead of some common one that might not like being shutdown

Reply:
That class is more for use in the openHAB core, not necessarily the bindings. Thread pools managed by ThreadPoolManager live outside the standard handler lifecycle which means that they are a poor choice to use if you want a threadpool that is only dedicated for a specific handler. Many bindings manage their thread pools, so there is nothing unusual about doing this.

If you look at the ThreadPoolManager source code you will notice that it does not recreate a threadpool if it is shutdown. That means that if a handler is restarted it will shutdown the threadpool and then restart the handler with a dead threadpool. Since we need to be able to shutdown the threadpool we have no choice other than to manage the threadpool instance ourselves.

Reply:
cool, will implement it like this than and will do some more testing